packages/grid/x-data-grid-pro/src/hooks/features/columnReorder/useGridColumnReorder.tsx

```diff
@@ -97,6 +98,69 @@ export const useGridColumnReorder = (
       });
 
       originColumnIndex.current = apiRef.current.getColumnIndex(params.field, false);
+
+      const draggingColumnGroupPath = apiRef.current.unstable_getColumnGroupPath(params.field);
+
+      const visibleColumnIndex = apiRef.current.getColumnIndex(params.field, true);
+      const visibleColumns = apiRef.current.getVisibleColumns();
+      const groupsLookup = apiRef.current.unstable_getAllGroupDetails();
+
+      // The limitingGroupId is the id of the group from which the dragged column should not escape
+      let limitingGroupId: string | null = null;
+
+      draggingColumnGroupPath.forEach((groupId) => {
+        if (!groupsLookup[groupId]?.freeReordering) {
+          // Only consider group that are made of more than one column
+          if (
+            visibleColumnIndex > 0 &&
+            visibleColumns[visibleColumnIndex - 1].groupPath?.includes(groupId)
+          ) {
+            limitingGroupId = groupId;
+          } else if (
+            visibleColumnIndex + 1 < visibleColumns.length &&
+            visibleColumns[visibleColumnIndex + 1].groupPath?.includes(groupId)
+          ) {
+            limitingGroupId = groupId;
+          }
+        }
+      });
+
+      forbiddenIndexes.current = {};
+
+      for (let indexToForbid = 0; indexToForbid < visibleColumns.length; indexToForbid += 1) {
+        const leftIndex = indexToForbid <= visibleColumnIndex ? indexToForbid - 1 : indexToForbid;
+        const rightIndex = indexToForbid < visibleColumnIndex ? indexToForbid : indexToForbid + 1;
+
+        if (limitingGroupId !== null) {
+          // verify this indexToForbid will be linked to the limiting group. Otherwise forbid it
+          let allowIndex = false;
+          if (leftIndex >= 0 && visibleColumns[leftIndex].groupPath?.includes(limitingGroupId)) {
+            allowIndex = true;
+          } else if (
+            rightIndex < visibleColumns.length &&
+            visibleColumns[rightIndex].groupPath?.includes(limitingGroupId)
+          ) {
+            allowIndex = true;
+          }
+          if (!allowIndex) {
+            forbiddenIndexes.current[indexToForbid] = true;
```

Comment:
If there's a hidden column, I can't reorder the first column of a group. Demo: https://codesandbox.io/s/basicgroupingdemo-demo-mui-x-forked-m5608x?file=/demo.tsx

![msedge_nVrMTLsVkT](https://user-images.githubusercontent.com/42154031/183669533-8bb08866-8a21-47eb-b43e-1c08395cb04e.gif)

I didn't read the full code but from my investigation it led to `forbiddenIndexes.current[targetColIndex]` being `true`. You need to consider that `targetColIndex` is related to all columns, not only the visible ones.

Reply:
Yes, I did not consider hidden columns. Should add test and correct this

Reply:
I fixed it by doing the computation on all the columns. not only the visible ones.

This add some edge cases to handle. I did the following one:

A and B  are in the same group but B is hidden. When I drop a column on A, it should not be placed on the right of A (it would break the group) but on the right of B.

A second edge case I did not solved is when A and B are in the same group and B is ignored. Curently you can not drag A because it would break grouping. But we could reorder all the hidden columns that are in the same group.

For the second point I prefer to do it when implementing the group reorder